cps/kni.c

```diff
@@ -294,6 +298,295 @@ kni_config(struct rte_kni *kni, struct gatekeeper_if *iface)
+		}
+	}
+	if (tb[RTA_SRC]) {
+		if (family == AF_INET6) {
+			struct in6_addr *addr =
+				mnl_attr_get_payload(tb[RTA_SRC]);
+			if (cps_conf->debug)
+				inet6_print("src", addr);
+		} else {
+			struct in_addr *addr =
+				mnl_attr_get_payload(tb[RTA_SRC]);
+			if (cps_conf->debug)
+				inet_print("src", addr);
+		}
+	}
+	if (tb[RTA_OIF]) {
+		if (cps_conf->debug)
+			RTE_LOG(INFO, GATEKEEPER, "cps update: oif=%u\n",
+				mnl_attr_get_u32(tb[RTA_OIF]));
+	}
+	if (tb[RTA_FLOW]) {
+		if (cps_conf->debug)
+			RTE_LOG(INFO, GATEKEEPER, "cps update: flow=%u\n",
+				mnl_attr_get_u32(tb[RTA_FLOW]));
+	}
+	if (tb[RTA_PREFSRC]) {
+		if (family == AF_INET6) {
+			struct in6_addr *addr =
+				mnl_attr_get_payload(tb[RTA_PREFSRC]);
+			if (cps_conf->debug)
+				inet6_print("prefsrc", addr);
+		} else {
+			struct in_addr *addr =
+				mnl_attr_get_payload(tb[RTA_PREFSRC]);
+			if (cps_conf->debug)
+				inet_print("prefsrc", addr);
+		}
+	}
+	if (tb[RTA_GATEWAY]) {
+		if (family == AF_INET6) {
+			struct in6_addr *addr =
+				mnl_attr_get_payload(tb[RTA_GATEWAY]);
+			if (cps_conf->debug)
+				inet6_print("gw", addr);
+		} else {
+			struct in_addr *addr =
+				mnl_attr_get_payload(tb[RTA_GATEWAY]);
+			if (cps_conf->debug)
+				inet_print("gw", addr);
+		}
+	}
+	if (tb[RTA_PRIORITY]) {
+		if (cps_conf->debug)
+			RTE_LOG(INFO, GATEKEEPER, "cps update: prio=%u\n",
+				mnl_attr_get_u32(tb[RTA_PRIORITY]));
+	}
+}
+
+static int
+data_ipv4_attr_cb(const struct nlattr *attr, void *data)
+{
+	const struct nlattr **tb = data;
+	int type = mnl_attr_get_type(attr);
+
+	/* Skip unsupported attribute in user-space. */
+	if (mnl_attr_type_valid(attr, RTA_MAX) < 0)
+		return MNL_CB_OK;
+
+	switch(type) {
+	case RTA_MULTIPATH:
+		if (mnl_attr_validate(attr, MNL_TYPE_NESTED) < 0)
+			return MNL_CB_ERROR;
+		break;
+	case RTA_TABLE:
+	case RTA_DST:
+	case RTA_SRC:
+	case RTA_OIF:
+	case RTA_FLOW:
+	case RTA_PREFSRC:
+	case RTA_GATEWAY:
+	case RTA_PRIORITY:
+		if (mnl_attr_validate(attr, MNL_TYPE_U32) < 0)
+			return MNL_CB_ERROR;
+		break;
+	}
+	tb[type] = attr;
+	return MNL_CB_OK;
+}
+
+static int
+data_ipv6_attr_cb(const struct nlattr *attr, void *data)
+{
+	const struct nlattr **tb = data;
+	int type = mnl_attr_get_type(attr);
+
+	/* Skip unsupported attribute in user-space. */
+	if (mnl_attr_type_valid(attr, RTA_MAX) < 0)
+		return MNL_CB_OK;
+
+	switch(type) {
+	case RTA_MULTIPATH:
+		if (mnl_attr_validate(attr, MNL_TYPE_NESTED) < 0)
+			return MNL_CB_ERROR;
+		break;
+	case RTA_TABLE:
+	case RTA_OIF:
+	case RTA_FLOW:
+	case RTA_PRIORITY:
+		if (mnl_attr_validate(attr, MNL_TYPE_U32) < 0)
+			return MNL_CB_ERROR;
+		break;
+	case RTA_DST:
+	case RTA_SRC:
+	case RTA_PREFSRC:
+	case RTA_GATEWAY:
+		if (mnl_attr_validate2(attr, MNL_TYPE_BINARY,
+					sizeof(struct in6_addr)) < 0)
+			return MNL_CB_ERROR;
+		break;
+	}
+	tb[type] = attr;
+	return MNL_CB_OK;
+}
+
+static int
+route_cb(const struct nlmsghdr *nlh, void *arg)
+{
+	struct nlattr *tb[RTA_MAX + 1] = {};
```

Comment:
What about using `__RTA_MAX` instead of `RTA_MAX`? This is a style comment, so feel free to ignore it if you don't agree with it. It's just that the `+ 1` distracts while reviewing the code.